### ISSUE TITLE
Cannot build dockerfile tools/dockerfile/interoptest/grpc_interop_android_java

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#11771]
`docker build -t temp tools/dockerfile/interoptest/grpc_interop_android_java`

```
...
Download https://jcenter.bintray.com/junit/junit/4.12/junit-4.12.jar
Download https://jcenter.bintray.com/javax/annotation/javax.annotation-api/1.2/javax.annotation-api-1.2.jar
The setTestClassesDir(File) method has been deprecated and is scheduled to be removed in Gradle 5.0. Please use the setTestClassesDirs(FileCollection) method instead.
The getTestClassesDir() method has been deprecated and is scheduled to be removed in Gradle 5.0. Please use the getTestClassesDirs() method instead.
The ConfigurableReport.setDestination(Object) method has been deprecated and is scheduled to be removed in Gradle 5.0. Please use the method ConfigurableReport.setDestination(File) instead.

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':app:prepareDebugDependencies'.
> Could not resolve all task dependencies for configuration ':app:_debugApk'.
   > Could not find com.android.support:appcompat-v7:22.2.1.
     Searched in the following locations:
         file:/grpc-java/android-interop-testing/.android/extras/m2repository/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.pom
         file:/grpc-java/android-interop-testing/.android/extras/m2repository/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.jar
         file:/grpc-java/android-interop-testing/.android/extras/google/m2repository/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.pom
         file:/grpc-java/android-interop-testing/.android/extras/google/m2repository/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.jar
         file:/grpc-java/android-interop-testing/.android/extras/android/m2repository/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.pom
         file:/grpc-java/android-interop-testing/.android/extras/android/m2repository/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.jar
         file:/root/.m2/repository/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.pom
         file:/root/.m2/repository/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.jar
         https://jcenter.bintray.com/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.pom
         https://jcenter.bintray.com/com/android/support/appcompat-v7/22.2.1/appcompat-v7-22.2.1.jar
     Required by:
         project :app
   > Could not find com.android.support:support-annotations:23.1.1.
     Searched in the following locations:
         file:/grpc-java/android-interop-testing/.android/extras/m2repository/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.pom
         file:/grpc-java/android-interop-testing/.android/extras/m2repository/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.jar
         file:/grpc-java/android-interop-testing/.android/extras/google/m2repository/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.pom
         file:/grpc-java/android-interop-testing/.android/extras/google/m2repository/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.jar
         file:/grpc-java/android-interop-testing/.android/extras/android/m2repository/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.pom
         file:/grpc-java/android-interop-testing/.android/extras/android/m2repository/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.jar
         file:/root/.m2/repository/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.pom
         file:/root/.m2/repository/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.jar
         https://jcenter.bintray.com/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.pom
         https://jcenter.bintray.com/com/android/support/support-annotations/23.1.1/support-annotations-23.1.1.jar
     Required by:
         project :app
   > Could not find com.google.android.gms:play-services-base:7.3.0.
     Searched in the following locations:
         file:/grpc-java/android-interop-testing/.android/extras/m2repository/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.pom
         file:/grpc-java/android-interop-testing/.android/extras/m2repository/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.jar
         file:/grpc-java/android-interop-testing/.android/extras/google/m2repository/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.pom
         file:/grpc-java/android-interop-testing/.android/extras/google/m2repository/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.jar
         file:/grpc-java/android-interop-testing/.android/extras/android/m2repository/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.pom
         file:/grpc-java/android-interop-testing/.android/extras/android/m2repository/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.jar
         file:/root/.m2/repository/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.pom
         file:/root/.m2/repository/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.jar
         https://jcenter.bintray.com/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.pom
         https://jcenter.bintray.com/com/google/android/gms/play-services-base/7.3.0/play-services-base-7.3.0.jar
     Required by:
         project :app

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED in 38s
```

Added in https://github.com/grpc/grpc/pull/11662

**Comment by jtattermusch** [comment]
CC @murgatroid99

**Comment by ericgribkoff** [comment]
The `appcompat-v7` and `support-annotations` dependencies are addressed by https://github.com/grpc/grpc-java/pull/3227. The failure to resolve `play-services-base` is a little more troublesome: if I change the Dockerfile to do `../assembleDebug || ../assembleDebug` it works, i.e., Gradle successfully downloads the dependencies on the _second_ try. 

@jtattermusch  If we need a quick fix, I can send a PR to hack the dockerfile to fix this, but a better solution is probably going to be directly installing the Android SDKs on the docker image, rather than the current license "trick" to get Gradle to do this automatically. https://grpc-testing.appspot.com/job/gRPC%20Android%20Firebase/ should be fine, as the cached docker image already has the necessary SDKs.

**Comment by jtattermusch** [comment]
@ericgribkoff  I took another look at the Dockerfile and it looks like it's fundamentally broken in the sense that it doesn't follow the way we work with interop docker images. Our dockerfiles are considered "base" docker images and they never touch the actual grpc code.
The idea is you have a "base" docker image with all the dependencies and you only bring in the grpc version being tested once you actually want to run the test.  (hence the build_interop.sh scripts for all other dockerfiles: 
https://github.com/grpc/grpc/blob/master/tools/dockerfile/interoptest/grpc_interop_csharp/build_interop.sh)

So, your work in #11662 will probably need a bit of restructuring (sadly that wasn't caught in the review).

Here's the snipped where you pull java in the Dockerfile (and you shouldn't)
```
# Download and install grpc-java
--
  |   |   | +  WORKDIR /
  |   |   | +  RUN git clone https://github.com/grpc/grpc-java.git
  |   |   | +  WORKDIR /grpc-java
  |   |   | +  RUN ./gradlew install
```

Besides all this, I suspect  `RUN git clone https://github.com/grpc/grpc-java.git` will only run once  and then will be cached by docker and never run again (so you'll be testing against old version of grpc-java without knowing.

One immediate problem with you dockerfile is that it break our push_images script:
https://github.com/grpc/grpc/blob/master/tools/dockerfile/push_testing_images.sh (which is why I originally reported this issue).

**Comment by ericgribkoff** [comment]
> Besides all this, I suspect RUN git clone https://github.com/grpc/grpc-java.git will only run once and then will be cached by docker and never run again (so you'll be testing against old version of grpc-java without knowing.

No, the actual test downloads and rebuilds gRPC Java each time: https://github.com/grpc/grpc/blob/master/tools/run_tests/interop/android/run_android_tests_on_firebase.sh.

The Dockerfile's building of gRPC Java is essentially a sentinel for deciding if the Android dependencies are correctly set up on the image. It can be removed from the Dockerfile, but obviously also served its purpose in this case, as it caught a failure based on the new dependency resolution for gradle. 

>  I took another look at the Dockerfile and it looks like it's fundamentally broken in the sense that it doesn't follow the way we work with interop docker images. Our dockerfiles are considered "base" docker images and they never touch the actual grpc code.

"Fundamentally broken" seems a bit strong here. Nothing like this is mentioned in the comments for `push_testing_images.sh`. This custom is documented where?

> One immediate problem with you dockerfile is that it break our push_images script:
> https://github.com/grpc/grpc/blob/master/tools/dockerfile/push_testing_images.sh (which is why I originally reported this issue).

As mentioned above, the building of gRPC Java in the Dockerfile can be removed. Doing so right now will break the test on Jenkins until I figure out how to automatically resolve the `play-services-base` dependency, but if it's an immediate need I can do so. Or, preferably, I can send a PR (temporarily?) excluding the android interop from the files picked up by push_testing_images.sh.

**Comment by jtattermusch** [comment]
On Thu, Jul 13, 2017 at 5:20 PM, Eric Gribkoff <notifications@github.com>
wrote:

> Besides all this, I suspect RUN git clone https://github.com/grpc/grpc-
> java.git will only run once and then will be cached by docker and never
> run again (so you'll be testing against old version of grpc-java without
> knowing.
>
> No, the actual test downloads and rebuilds gRPC Java each time:
> https://github.com/grpc/grpc/blob/master/tools/run_tests/
> interop/android/run_android_tests_on_firebase.sh.
>
The Dockerfile's building of gRPC Java is essentially a sentinel for
> deciding if the Android dependencies are correctly set up on the image. It
> can be removed from the Dockerfile,
>
but obviously also served its purpose in this case, as it caught a failure
> based on the new dependency resolution for gradle.
>
Well, the same failure would be caught if you tried to run the test,
wouldn't it? I would argue that when it should fail.

Also, pre-downloading and installing some ("latest version at the time of
caching" - so essentially random) version of grpc-java seems error-prone in
the sense that it makes it possible to accidentally running the tests using
the old version of grpc-java (and it could mask bugs introduced in later
versions).

also, IF we really need to pre-download java (and I don't think we do), we
should at least use a pinned version to avoid breaking the dockerfile build
in the future (and add a comment that we are installing grpc-java and that
it's not THE grpc-java installation that will get actually tested).



> I took another look at the Dockerfile and it looks like it's fundamentally
> broken in the sense that it doesn't follow the way we work with interop
> docker images. Our dockerfiles are considered "base" docker images and they
> never touch the actual grpc code.
>
> "Fundamentally broken" seems a bit strong here. Nothing like this is
> mentioned in the comments for push_testing_images.sh. This custom is
> documented where?
>
Perhaps I used a too strong expression - sorry for that - but yeah, I think
it is currently broken.
Unfortunately, it's not documented properly. run_interop_tests.py kind of
implies this custom by the way it works. Perhaps we could add a README.md
to tools/dockerfile that explains how we treat the dockerfiles. I am hoping
in the future this will
get better because of the OWNERS files we've setup (so important changes
will make it the right reviewers).


> One immediate problem with you dockerfile is that it break our push_images
> script:
> https://github.com/grpc/grpc/blob/master/tools/dockerfile/
> push_testing_images.sh (which is why I originally reported this issue).
>
> As mentioned above, the building of gRPC Java in the Dockerfile can be
> removed. Doing so right now will break the test on Jenkins until I figure
> out how to automatically resolve the play-services-base dependency, but
> if it's an immediate need I can do so. Or, preferably, I can send a PR
> (temporarily?) excluding the android interop from the files picked up by
> push_testing_images.sh.
>

This issue is not super urgent, but we should resolve it soon enough.
Suggestions:

1) merge  grpc/grpc-java#3227 <https://github.com/grpc/grpc-java/pull/3227>
2) use ../assembleDebug || ../assembleDebug workaround in run_android_
tests_on_firebase.sh
<https://github.com/grpc/grpc/blob/master/tools/run_tests/interop/android/run_android_tests_on_firebase.sh>
with
a TODO to address it (rather than doing the preparation in the base
Dockerfile)

**Comment by ericgribkoff** [comment]
Fixed via #11916
